core/src/main/scala/flow/Generator.scala

```diff
@@ -95,10 +95,15 @@ class Generator[+T, +Out](private[flow] val source: () ⇒ Future[Source[T, Futu
   def flatMapMaterializedValue[Out2](f: Out ⇒ Future[Out2])(implicit ec: ExecutionContext) =
     use(() ⇒ source().map(_.mapMaterializedValue(outFuture ⇒ outFuture.flatMap(f))))
 
-  def recoverWith[U >: T, Out2 >: Out](p: PartialFunction[Throwable, Future[Source[U, Future[Out2]]]])(
+  def recoverWithRetries[U >: T, Out2 >: Out](attempts: Int, p: PartialFunction[Throwable, Source[U, akka.NotUsed]])(
       implicit ec: ExecutionContext): Generator[U, Out2] =
-    use(() ⇒ source().recoverWith(p))
+    use(() ⇒ source().map(_.recoverWithRetries(attempts, p)))
 
+  //allows a recoverWith to be applied to instantiation of the source (which is a Future)
+  def recoverSourceWith[U >: T, Out2 >: Out](p: PartialFunction[Throwable, Future[Source[U, Future[Out2]]]])(
+      implicit ec: ExecutionContext): Generator[U, Out2] =
+    use(() ⇒ source().recoverWith(p))
```

Comment:
Does it make sense to also define a `recoverWith` that does both of these things? I think that usually, that's what you want to do (i.e., you don't really know/care whether the future is failing or whether the source is failing)

Reply:
Maybe?  They have slightly different outputs because in one case you can return a whole generator and in the other case just a source.